### PR TITLE
Move GeneralPredicates logic to kubelet.

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind cleanup
/sig node
/sig scheduling

**What this PR does / why we need it**:

Background: kubelet reuses some scheduler code (e.g., GeneralPredicates) to check if a Pod fits on a node. However, along with recent code cleanup in scheduler side, scheduler doesn't consume GeneralPredicates anymore, instead, it decoupled the fitting behavior into individual filter plugins.

This is PR does nothing but move GeneralPredicates (and its related types) to kubelet.

**Which issue(s) this PR fixes**:

Part of #85822.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/lifecycle/predicate.go [hunk 1]
```diff
@@ -168,3 +172,68 @@ func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulernodeinfo.Nod
...
+	return fmt.Sprintf("Predicate %s failed", e.PredicateName)
+}
+
+// GetReason returns the reason of the PredicateFailureError.
+func (e *PredicateFailureError) GetReason() string {
+	return e.PredicateDesc
+}
+
+// GeneralPredicates checks a group of predicates that the kubelet cares about.
+func GeneralPredicates(pod *v1.Pod, _ interface{}, nodeInfo *schedulernodeinfo.NodeInfo) (bool, []PredicateFailureReason, error) {
```
Comment (wgliang) [inline pkg/kubelet/lifecycle/predicate.go]: An early question: Is the filter item configurable (in some cases, don't want pods to be evicted).

File: pkg/kubelet/lifecycle/predicate.go [hunk 2]
```diff
@@ -126,11 +130,11 @@ func (w *predicateAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult
 		// If there are failed predicates, we only return the first one as a reason.
 		r := reasons[0]
 		switch re := r.(type) {
-		case *predicates.PredicateFailureError:
+		case *PredicateFailureError:
```
Comment (haosdent) [inline pkg/kubelet/lifecycle/predicate.go]: Would this cause problems during kubernetes upgrade? For example, if update a kubernetes cluste from 1.17 to 1.18. Some kubelet are still using the 1.17 version while the kube-scheduler is using 1.18.

File: pkg/kubelet/lifecycle/predicate.go [hunk 3]
```diff
@@ -168,3 +172,68 @@ func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulernodeinfo.Nod
 	}
 	return podCopy
 }
+
+// InsufficientResourceError is an error type that indicates what kind of resource limit is
+// hit and caused the unfitting failure.
+type InsufficientResourceError struct {
```
Comment (haosdent) [inline pkg/kubelet/lifecycle/predicate.go]: What if we just use `noderesources.InsufficientResource`

File: pkg/kubelet/lifecycle/predicate.go [hunk 4]
```diff
@@ -168,3 +172,68 @@ func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulernodeinfo.Nod
...
+	noderesources.InsufficientResource
+}
+
+func (e *InsufficientResourceError) Error() string {
+	return fmt.Sprintf("Node didn't have enough resource: %s, requested: %d, used: %d, capacity: %d",
+		e.ResourceName, e.Requested, e.Used, e.Capacity)
+}
+
+// PredicateFailureReason interface represents the failure reason of a predicate.
+type PredicateFailureReason interface {
```
Comment (haosdent) [inline pkg/kubelet/lifecycle/predicate.go]: Do we still need this or can remove it and update the interface of admission handler to 

```
func (n *AdmissionFailureHandlerStub) HandleAdmissionFailure(admitPod *v1.Pod, failureReasons []string) (bool, []string, error) {
```

File: pkg/kubelet/lifecycle/predicate.go [hunk 5]
```diff
@@ -168,3 +172,68 @@ func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulernodeinfo.Nod
...
+	return fmt.Sprintf("Insufficient %v", e.ResourceName)
+}
+
+// GetInsufficientAmount returns the amount of the insufficient resource of the error.
+func (e *InsufficientResourceError) GetInsufficientAmount() int64 {
+	return e.Requested - (e.Capacity - e.Used)
+}
+
+// PredicateFailureError describes a failure error of predicate.
+type PredicateFailureError struct {
```
Comment (haosdent) [inline pkg/kubelet/lifecycle/predicate.go]: How about change it to private `predicateFailureError`

File: pkg/kubelet/lifecycle/admission_failure_handler_stub.go [hunk 6]
```diff
@@ -31,6 +30,6 @@ func NewAdmissionFailureHandlerStub() *AdmissionFailureHandlerStub {
 	return &AdmissionFailureHandlerStub{}
 }
 
-func (n *AdmissionFailureHandlerStub) HandleAdmissionFailure(admitPod *v1.Pod, failureReasons []predicates.PredicateFailureReason) (bool, []predicates.PredicateFailureReason, error) {
+func (n *AdmissionFailureHandlerStub) HandleAdmissionFailure(admitPod *v1.Pod, failureReasons []PredicateFailureReason) (bool, []PredicateFailureReason, error) {
```
Comment (haosdent) [inline pkg/kubelet/lifecycle/admission_failure_handler_stub.go]: Would it cause compatibility issues if we update the interface here? Would someone writer their custom admission failure handler which out of the repo?

File: pkg/kubelet/lifecycle/predicate.go (hunk 1 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/predicate.go]: Probably. If external users want to vendor some stateless filtering functions. But keep in mind that not all filter plugin can function well without pre-calculating (pre-filtering).

File: pkg/kubelet/lifecycle/predicate.go (hunk 2 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/predicate.go]: Nope. This was a library-level dependency, not runtime dependency (like API). The dependency was sealed at the moment when the old-version kubelet binary was cut.

File: pkg/kubelet/lifecycle/predicate.go (hunk 3 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/predicate.go]: Given that scheduler don't use `InsufficientResource` to output verbosed log, or rely on the specific shortage amount to make other scheduling decisions, it sounds unnecessary to me.

File: pkg/kubelet/lifecycle/predicate.go (hunk 4 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/predicate.go]: We'd better not. This interface have multiple implementations which are all in use:

- PredicateFailureError
- InsufficientResourceError

File: pkg/kubelet/lifecycle/predicate.go (hunk 5 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/predicate.go]: Technically yes, but the other error type "InsufficientResourceError" is used in other packages, so IMO it's fine to leave them both as public.

File: pkg/kubelet/lifecycle/admission_failure_handler_stub.go (hunk 6 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/admission_failure_handler_stub.go]: Nope. Explained above.

File: pkg/kubelet/lifecycle/predicate.go (hunk 1 above)
Comment (ahg-g) [inline pkg/kubelet/lifecycle/predicate.go]: remove the ```_ interface``` parameter. I would also remove the ```bool``` return value, it is redundant (the caller can ```len(reasons) == 0``` to check if the pod fits the node or not).

**Review by ahg-g — COMMENTED** [review]
Thanks, just a nit.

File: pkg/kubelet/lifecycle/predicate.go (hunk 1 above)
Comment (Huang-Wei) [reply pkg/kubelet/lifecycle/predicate.go]: done.

**Comment by Huang-Wei** [comment]
/hold
for squashing commits later.

**Comment by ahg-g** [comment]
/lgtm

For the scheduler

**Comment by haosdent** [comment]
```
pkg/controller/.import-restrictions:        "k8s.io/kubernetes/pkg/scheduler/algorithm/predicates",
```
Seems this file forgot to update

**Comment by Huang-Wei** [comment]
> Seems this file forgot to update

Given that you've opened #87406, I'm going to leave "k8s.io/kubernetes/pkg/scheduler/algorithm/predicates" to you in that PR, is that OK?

**Comment by Huang-Wei** [comment]
@kubernetes/sig-node-pr-reviews for approval.

**Comment by dashpole** [comment]
/lgtm
for node changes

**Comment by Huang-Wei** [comment]
Thanks @dashpole .

@dchen1107 @derekwaynecarr Could you kindly /approve? The background and reason are listed in the description section.

**Pull request merged**